packages/compiler/test/server/test-colorization.ts

```diff
@@ -0,0 +1,776 @@
+import { createTestServerHost } from "../../testing/test-server-host.js";
+
+const { parseRawGrammar, Registry } = vscode_textmate;
+const { createOnigScanner, createOnigString, loadWASM } = vscode_oniguruma;
+
+export type MetaScope = `meta.${string}.cadl`;
+export type TokenScope = CadlScope | MetaScope;
+
+interface Token {
+  text: string;
+  scope: TokenScope;
+}
+
+function createToken(text: string, scope: TokenScope): Token {
+  return { text, scope };
+}
+
+type Tokenize = (input: string) => Promise<Token[]>;
+
+const Token = {
+  keywords: {
+    model: createToken("model", "keyword.other.cadl"),
+    operation: createToken("op", "keyword.other.cadl"),
+    namespace: createToken("namespace", "keyword.other.cadl"),
+    interface: createToken("interface", "keyword.other.cadl"),
+    alias: createToken("alias", "keyword.other.cadl"),
+    extends: createToken("extends", "keyword.other.cadl"),
+    is: createToken("is", "keyword.other.cadl"),
+    other: (text: string) => createToken(text, "keyword.other.cadl"),
+  },
+
+  meta: (text: string, meta: string) => createToken(text, `meta.${meta}.cadl`),
+
+  identifiers: {
+    variable: (name: string) => createToken(name, "variable.name.cadl"),
+    functionName: (name: string) => createToken(name, "entity.name.function.cadl"),
+    tag: (name: string) => createToken(name, "entity.name.tag.cadl"),
+    type: (name: string) => createToken(name, "entity.name.type.cadl"),
+  },
+
+  operators: {
+    assignment: createToken("=", "keyword.operator.assignment.cadl"),
+    optional: createToken("?", "keyword.operator.optional.cadl"),
+    typeAnnotation: createToken(":", "keyword.operator.type.annotation.cadl"),
+    spread: createToken("...", "keyword.operator.spread.cadl"),
+  },
+
+  punctuation: {
+    comma: createToken(",", "punctuation.comma.cadl"),
+    accessor: createToken(".", "punctuation.accessor.cadl"),
+    openBracket: createToken("[", "punctuation.squarebracket.open.cadl"),
+    closeBracket: createToken("]", "punctuation.squarebracket.close.cadl"),
+    openBrace: createToken("{", "punctuation.curlybrace.open.cadl"),
+    closeBrace: createToken("}", "punctuation.curlybrace.close.cadl"),
+    openParen: createToken("(", "punctuation.parenthesis.open.cadl"),
+    closeParen: createToken(")", "punctuation.parenthesis.close.cadl"),
+    semicolon: createToken(";", "punctuation.terminator.statement.cadl"),
+
+    typeParameters: {
+      begin: createToken("<", "punctuation.definition.typeparameters.begin.cadl"),
+      end: createToken(">", "punctuation.definition.typeparameters.end.cadl"),
+    },
+  },
+
+  literals: {
+    numeric: (text: string) => createToken(text, "constant.numeric.cadl"),
+    string: (text: string) =>
+      createToken(text.startsWith('"') ? text : '"' + text + '"', "string.quoted.double.cadl"),
+  },
+} as const;
+
+testColorization("semantic colorization", tokenizeSemantic);
+testColorization("tmlanguage", tokenizeTMLanguage);
```

Comment:
So the semantic coloring basically changes how we resolve the color token but in the end it still create the same token.

Reply:
It has a bit more detail available already, but that's collapsed in tokenizeSemantic to map to tokenizeTMLanguage. In practice, the bits of more detail (for example, this is a namespace, and this is an interface vs. just "type" don't actually change the colors in any VS Code theme I've tried. Eventually, we probably want to test this precision better, but my main goal here was parity with tmlanguage to make the playground pretty.

Reply:
Arguably, there isn't much value yet in calling this code in VS Code or other editors with tmlanguage support, but I'm leaving it in so that we get some dogfooding coverage and people will see if there are problems with this in more places.